src/main/scala/dxWDL/CompilerTopologicalSort.scala

```diff
@@ -0,0 +1,265 @@
+/**
+  *  Preprocessing pass, simplify the original WDL.
+  *
+  *  Instead of handling expressions in calls directly,
+  *  we lift the expressions, generate auxiliary variables, and
+  *  call the task with values or variables (no expressions).
+  *
+  *  A difficulty we face here, is avoiding using internal
+  *  representations used by wdl4s. For example, we want to reorganize
+  *  scatter blocks, however, we cannot create valid new wdl4s scatter
+  *  blocks. Instead, we pretty print a new workflow and then load it.
+  */
+package dxWDL
+
+import java.io.{File, FileWriter, PrintWriter}
+import java.nio.file.{Path, Paths}
+import scala.collection.mutable.Queue
+import wdl4s.AstTools
+import wdl4s.AstTools.EnhancedAstNode
+import wdl4s.{Call, Declaration, Scatter, Scope,
+    Task, TaskCall, TaskOutput,
+    WdlExpression, WdlNamespace, WdlNamespaceWithWorkflow,
+    Workflow, WorkflowCall, WdlSource, GraphNode}
+import wdl4s.command.{ParameterCommandPart, StringCommandPart}
+import wdl4s.parser.WdlParser.{Ast, AstNode, Terminal}
+import wdl4s.types._
+import wdl4s.values._
+import wdl4s.WdlExpression.AstForExpressions
+
+object CompilerTopologicalSort {
+    // TODO: move common procs to a common place
+
+    // Add a suffix to a filename, before the regular suffix. For example:
+    //  xxx.wdl -> xxx.sorted.wdl
+    def addFilenameSuffix(src: Path, secondSuffix: String) : String = {
+        val fName = src.toFile().getName()
+        val index = fName.lastIndexOf('.')
+        if (index == -1) {
+            fName + secondSuffix
+        }
+        else {
+            val prefix = fName.substring(0, index)
+            val suffix = fName.substring(index)
+            prefix + secondSuffix + suffix
+        }
+    }
+
+    case class State(cef: CompilerErrorFormatter,
+                     terminalMap: Map[Terminal, WdlSource],
+                     verbose: Boolean)
+
+
+    // Generic topological sorting procedure
+    // Scala code only slightly modified from https://gist.github.com/ThiporKong/4399695
+    def tsort[A](edges: Traversable[(A, A)]): Iterable[A] = {
+        def tsort(toPreds: Map[A, Set[A]], done: Iterable[A]): Iterable[A] = {
+            // Partition the set of nodes into those that have a predecessor and those that do not
+            // (Actually this will partition the map into two)
+            val (noPreds, hasPreds) = toPreds.partition { keyval =>
+                val predSet = keyval._2
+                predSet.isEmpty
+            }
+
+            // If there are no nodes left with no predecessors, either we are done recursing or there is a cycle
+            if (noPreds.isEmpty) {
+
```

Comment:
Personally, I prefer to minimize the number of empty lines. The issue is that methods become very long, violating the rule that they should fit into one editor page.